packages/vm/src/imports.rs

```diff
@@ -138,9 +187,19 @@ fn write_to_contract<S: Storage, Q: Querier>(ctx: &mut Ctx, input: &[u8]) -> VmR
 pub fn do_query_chain<S: Storage, Q: Querier>(ctx: &mut Ctx, request_ptr: u32) -> VmResult<u32> {
     let request = read_region(ctx, request_ptr, MAX_LENGTH_QUERY_CHAIN_REQUEST)?;
 
-    let (res, used_gas) =
-        with_querier_from_context::<S, Q, _, _>(ctx, |querier| Ok(querier.raw_query(&request)?))?;
-    process_gas_info::<S, Q>(ctx, used_gas)?;
+    let ffi_result =
+        with_querier_from_context::<S, Q, _, _>(ctx, |querier| Ok(querier.raw_query(&request)))?;
+
+    let res = match ffi_result {
```

Comment:
With the comment above to change the return type to (Result<T, FfiError>, GasInfo), this would be something like:

```rust
let (res, gas_info) = with_querier_from_context::<S, Q, _, _>(ctx, |querier| Ok(querier.raw_query(&request)))?;
process_gas_info::<S, Q>(ctx, gas_info)?;
let res = res.map_err(VmError::from)?;
```

I think it would be simpler (even if the type does look a bit funny)

Reply:
Actually, given the comments above, maybe we can add a method to `FfiResult` to allow easy ?

```rust
let res = with_querier_from_context::<S, Q, _, _>(ctx, |querier| Ok(querier.raw_query(&request)))?;
let res = res.process::<S,Q>(ctx)?; 

fn process(self, ctx: &mut Context) -> Result<T, VmError> {
    let (res, gas_info) = self;
    process_gas_info::<S, Q>(ctx, gas_info)?;
    res.map_err(VmError::from)
}
```

This would effectively just change the naive `let res = res?` to `let res = res.process(ctx)?` while ensuring all gas is charged (and compiler erroring on trying `?` without charging the gas)

Reply:
This is interesting but you'd still have to process the gas in the `Ok` case. Could be interesting to have a method `to_vm_result` on `FfiResult` that performs the charging and conversion. Both approaches are only possible by defining an extention trait for FfiResult, because under the hood it's an alias for Result which is a foreign type (not implemented by this crate)

Reply:
Also this won't be used by humanize/canonicalize which have different needs

Reply:
Resolving this because parts are redundant with the other thread and parts are not possible to do